nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+	return c, nil
+}
+
+// Clear removes all the entries in the cache
+func (c *Cache) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(entries, c.Capacity)
+}
+
+// Purge removes the old elements in the cache
+func (c *Cache) Purge(now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.removeOldest(c.Capacity, now)
+}
+
+// Add adds a reply to the cache.
+func (c *Cache) Put(request *dns.Msg, reply *dns.Msg, proto dnsProtocol, flags uint8, now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		ent.setReply(reply, flags, now)
+	} else {
+		// If we will add a new item and the capacity has been exceeded, make some room...
+		if len(c.entries) >= c.Capacity {
+			c.removeOldest(1, now)
+		}
+		c.entries[key] = newCacheEntry(&question, reply, stResolved, now)
+	}
+}
+
+// Look up for a question's reply from the cache.
+// If no reply is stored in the cache, it returns a `nil` reply and no error. The caller can then `Wait()`
+// for another goroutine `Put`ing a reply in the cache.
+func (c *Cache) Get(request *dns.Msg, proto dnsProtocol, now time.Time) (reply *dns.Msg, err error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		reply, err = ent.getReply(request, now)
+		if ent.hasExpired(now) {
+			delete(c.entries, key)
+			reply = nil
+		}
+	} else {
+		// we are the first asking for this name: create an entry with no reply... the caller must wait
+		c.entries[key] = newCacheEntry(&question, nil, stPending, now)
+	}
+	return
+}
+
+// Wait for a reply for a question in the cache
+// Notice that the caller could Get() and then Wait() for a question, but the corresponding cache
+// entry could have been removed in between. In that case, the caller should retry the query (and
+// the user should increase the cache size!)
+func (c *Cache) Wait(request *dns.Msg, proto dnsProtocol, timeout time.Duration, now time.Time) (reply *dns.Msg, err error) {
+	// do not try to lock the cache: otherwise, no one else could `Put()` the reply
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if entry, found := c.entries[key]; found {
+		reply, err = entry.waitReply(request, timeout, now)
+	}
+	return
+}
+
+// Remove removes the provided question from the cache.
+func (c *Cache) Remove(question *dns.Question, proto dnsProtocol) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	key := cacheKey{*question, proto}
+	delete(c.entries, key)
+}
+
+// Len returns the number of entries in the cache.
+func (c *Cache) Len() int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return len(c.entries)
+}
+
+// removeOldest removes the oldest item(s) from the cache.
+// note: this method is not thread safe (it is a responsability of the caller function...)
+func (c *Cache) removeOldest(atLeast int, now time.Time) {
```

Comment:
(This would depend on Go's implementation of hash tables, but) I think a heap would make the most common operation (Get's) more expensive.

And, if we keep the hash table, the "sort them first" would imply the copy I do as, AFAIK, you cannot sort a hash table in Go. Then you can use the sorted data for either a) deleting entries in the hash table or b) re-inserting in the table...

Anyway, we are doing an early optimization... I would wait until this becomes a bottleneck. My guess is that new entries in the cache will come at the same rate as they expire their TTL, so we will never reach this expensive part of `removeOldest`...